File: sdk-java-blocking/src/test/java/dev/restate/sdk/blocking/AwakeableIdTest.java
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk.blocking;

import static dev.restate.sdk.core.impl.ProtoUtils.greetingResponse;

import dev.restate.sdk.core.CoreSerdes;
import dev.restate.sdk.core.impl.AwakeableIdTestSuite;
import dev.restate.sdk.core.impl.testservices.GreeterGrpc;
import dev.restate.sdk.core.impl.testservices.GreetingRequest;
import dev.restate.sdk.core.impl.testservices.GreetingResponse;
import io.grpc.BindableService;
import io.grpc.stub.StreamObserver;

public class AwakeableIdTest extends AwakeableIdTestSuite {

  private static class ReturnAwakeableId extends GreeterGrpc.GreeterImplBase
      implements RestateBlockingService {

    @Override
    public void greet(GreetingRequest request, StreamObserver<GreetingResponse> responseObserver) {
      String id = restateContext().awakeable(CoreSerdes.STRING_UTF8).id();
      responseObserver.onNext(greetingResponse(id));
      responseObserver.onCompleted();
    }
  }

  @Override
  protected BindableService returnAwakeableId() {
    return new ReturnAwakeableId();
  }
}

File: sdk-kotlin/src/test/kotlin/dev/restate/sdk/kotlin/AwakeableIdTest.kt
package dev.restate.sdk.kotlin

import dev.restate.sdk.core.TypeTag
import dev.restate.sdk.core.impl.AwakeableIdTestSuite
import dev.restate.sdk.core.impl.testservices.GreeterGrpcKt
import dev.restate.sdk.core.impl.testservices.GreetingRequest
import dev.restate.sdk.core.impl.testservices.GreetingResponse
import dev.restate.sdk.core.impl.testservices.greetingResponse
import io.grpc.BindableService
import kotlinx.coroutines.Dispatchers

internal class AwakeableIdTest : AwakeableIdTestSuite() {
  private class ReturnAwakeableId :
      GreeterGrpcKt.GreeterCoroutineImplBase(Dispatchers.Unconfined), RestateCoroutineService {

    override suspend fun greet(request: GreetingRequest): GreetingResponse {
      val id: String = restateContext().awakeable(TypeTag.STRING_UTF8).id
      return greetingResponse { message = id }
    }
  }

  override fun returnAwakeableId(): BindableService {
    return ReturnAwakeableId()
  }
}
